### PR TITLE
Add MLflow experiment tracking

### PR DESCRIPTION
**Pull request by Rasaboun**
## Summary

Adds MLflow as an experiment tracking option alongside W&B and UILogger.

- `MLflowLogger`: metrics, step-aligned images, checkpoint artifacts, training datasets
- `CompositeLogger`: run multiple loggers simultaneously (W&B + MLflow)
- LoRA model registration in MLflow Model Registry (versioning + lineage)
- `log_checkpoint` added to `EmptyLogger` and `UILogger` base interface
- Example config: `train_lora_flux_24gb_mlflow.yaml`

## Requirements

- **MLflow 3.x** — this PR uses the MLflow 3.x API (e.g. `mlflow.pyfunc.log_model(name=...)` instead of the deprecated `artifact_path` parameter). Tested on MLflow 3.10.1.

## Usage

```yaml
logging:
  use_mlflow: true
  project_name: "my-lora"
  mlflow_tracking_uri: "http://my-server:5000"
  mlflow_experiment_name: "flux-lora"
  mlflow_log_artifacts: true             # upload checkpoint .safetensors
  mlflow_register_model: true            # register final LoRA in Model Registry
  mlflow_registered_model_name: "my-lora" # optional: versioned name (null = log only)
```

Both W&B and MLflow can be enabled at the same time via `CompositeLogger`. Existing single-logger configs are unchanged.

## What gets logged

| Data | Where in MLflow | How |
|---|---|---|
| Training metrics (loss, lr) | Run → Metrics | `mlflow.log_metrics()` per step |
| Sample images | Run → Artifacts → `images/` | `mlflow.log_image()` step-aligned |
| Hyperparameters | Run → Parameters | Flattened config via `mlflow.log_params()` |
| Checkpoints | Run → Artifacts → `checkpoints/` | `mlflow.log_artifact()` (if `mlflow_log_artifacts: true`) |
| Training datasets | Run → Datasets | `mlflow.log_input()` with schema metadata |
| LoRA weights | Model Registry | `mlflow.pyfunc.log_model()` with stub (if `mlflow_register_model: true`) |

## Details

`MLflowLogger` follows the same `log()`/`commit()` two-phase pattern as `WandbLogger`. Metrics are batched per step via `mlflow.log_metrics()`. Images are logged as step-aligned artifacts using the `+` separator format matching upstream MLflow.

All MLflow API calls are wrapped in `try/except` so a tracking server failure never kills a training run.

`post_save_hook` now calls `log_checkpoint` to upload checkpoint artifacts. Since `file_path` gets reassigned to `optimizer.pt` later in `save()`, we snapshot it beforehand so the hook receives the actual checkpoint path.

### Model Registration

LoRA adapters are registered in the MLflow Model Registry using a minimal pyfunc stub. This provides versioning and lineage (which run produced which LoRA) without attempting inference — MLflow has no native diffusers adapter flavor yet ([mlflow/mlflow#22122](https://github.com/mlflow/mlflow/issues/22122)). Each training run with `mlflow_register_model: true` auto-increments the model version. The stub's `predict()` raises `NotImplementedError` with a clear message pointing users to load the weights via diffusers directly.

Model registration works for all architectures (sd1/2/3, sdxl, flux, pixart, auraflow, lumina2, qwen_image, zimage, wan21, etc.) — uses `model_config.arch` for type detection.

### Dataset Logging

Training datasets are logged at the start of each run via `mlflow.log_input()`. The dataset name (folder basename), source path, and schema (resolution, type, caption_ext, trigger_word) are recorded. Dataset values are also available in run parameters via the flattened config.

### DiffusionTrainer / UITrainer

`save()` in both subclasses now propagates the return value from `BaseSDTrainProcess.save()`, which returns the checkpoint path. This is needed for model registration to know which file to register.

## Test results

Tested end-to-end on **Modal A100** with FLUX Klein 4B (30 steps, MLflow 3.10.1):

- [x] Train a LoRA with `use_mlflow: true` — metrics (loss, lr) appear in MLflow UI at correct steps
- [x] Enable `mlflow_log_artifacts: true` — checkpoint .safetensors appear in Artifacts tab (3 checkpoints)
- [x] Sample images logged with step-aligned naming — 6 files in `images/` (PNG + compressed WebP)
- [x] Flattened hyperparameters logged as params (23 params)
- [x] Training config YAML logged as downloadable artifact
- [x] Verify existing W&B-only and UILogger-only configs are unaffected
- [x] Kill MLflow server mid-training — training continues without crash (error resilience confirmed)
- [x] Enable `mlflow_register_model: true` — model registered as version 1 in Model Registry
- [x] Set `mlflow_registered_model_name` and run twice — version auto-incremented to 2
- [x] Enable both `use_wandb` and `use_mlflow` — CompositeLogger dispatches to both (W&B synced 5 files + 3 media, MLflow got metrics + artifacts + model v3)

### Optional: reproducing on Modal GPU

<details>
<summary>Click to expand Modal test setup</summary>

The test infrastructure is not part of the PR but can be used to reproduce results.

**MLflow server on Modal** (`mlflow_server.py`):
```bash
modal deploy mlflow_server.py
```
Server flags needed behind Modal's reverse proxy:
- `--serve-artifacts` — proxy artifact uploads through the server
- `--default-artifact-root mlflow-artifacts:/mlflow/artifacts` — ensure reads also go through proxy
- `--allowed-hosts "*"` — Modal proxy changes Host headers

**Training runner** (`run_modal_mlflow_test.py`):
```bash
modal secret create huggingface HF_TOKEN=hf_xxx   # for gated models
modal run run_modal_mlflow_test.py
```

**Notes:**
- Clear output directory between runs to avoid checkpoint resumption (`start_step` = previous run's last step)
- Split pip install into layers (torch first, then requirements.txt, then mlflow) to avoid `resolution-too-deep`
- Use `Image.add_local_dir()` without `copy=True` for runtime mount (avoids image rebuild on code changes)
</details>

### CODE REVIEW

File: toolkit/logging_aitk.py [hunk 1]
```diff
@@ -83,6 +91,272 @@ def finish(self):
...
+        for lg in self._loggers:
+            lg.commit(step=step)
+
+    def log_image(self, *args, **kwargs):
+        for lg in self._loggers:
+            lg.log_image(*args, **kwargs)
+
+    def log_checkpoint(self, file_path: str):
+        for lg in self._loggers:
+            lg.log_checkpoint(file_path)
```
[the remark this answers is not in the document]
Comment (Rasaboun) [reply toolkit/logging_aitk.py]: Fixed in c99ff31 — added log_checkpoint no-op to UILogger
